File: hard/987.vertical-order-traversal-of-a-binary-tree.java
import java.util.ArrayList;
import java.util.HashMap;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode id=987 lang=java
 *
 * [987] Vertical Order Traversal of a Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, HashMap<Integer, List<Integer>>> colToRow = new HashMap<>(); // {col: {row : [val]} }
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        return myFirstSolution(root);
    }

    private List<List<Integer>> myFirstSolution(Tree root) {
        dfs(root, 0, 0);

        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        Integer[] cols = colToRow.keySet().toArray(new Integer[0]);
        Arrays.sort(cols);
        for (int col : cols) {
            HashMap<Integer, List<Integer>> rowToNodes = colToRow.get(col);
            ArrayList<Integer> ansVals = new ArrayList<>();
            Integer[] rows = rowToNodes.keySet().toArray(new Integer[0]);
            Arrays.sort(rows);
            for (int row : rows) {
                Integer[] valsInRow = rowToNodes.get(row).toArray(new Integer[0]);
                Arrays.sort(valsInRow);
                Collections.addAll(ansVals, valsInRow);
            }
            ans.add(ansVals);
        }

        return ans;
    }

    private void dfs(TreeNode root, int row, int col) {
        if (root == null) {
            return;
        }

        HashMap<Integer, List<Integer>> rowToNodes = null;
        if (colToRow.containsKey(col)) {
            rowToNodes = colToRow.get(col);
        } else {
            rowToNodes = new HashMap<Integer, List<Integer>>();
            colToRow.put(col, rowToNodes);
        }

        List<Integer> nodes = null;
        if (rowToNodes.containsKey(row)) {
            nodes = rowToNodes.get(row);
        } else {
            nodes = new ArrayList<Integer>();
            rowToNodes.put(row, nodes);
        }

        nodes.add(root.val);

        // System.out.println("val: " + root.val + " (" + row + ", " + col + ")");

        dfs(root.left, row + 1, col - 1);
        dfs(root.right, row + 1, col + 1);
    }
}
// @lc code=end

File: hard/987.vertical-order-traversal-of-a-binary-tree.js
/*
 * @lc app=leetcode id=987 lang=javascript
 *
 * [987] Vertical Order Traversal of a Binary Tree
 * [tree]
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
  const yMap = new Map(); // { y => { x => [val1, val2, ...] } }
  function dfs(root, x, y) {
    if (!root) {
      return null;
    }

    if (!yMap.has(y)) {
      yMap.set(y, new Map());
    }

    xMap = yMap.get(y);
    if (!xMap.has(x)) {
      xMap.set(x, []);
    }
    xMap.get(x).push(root.val);
    dfs(root.left, x + 1, y - 1);
    dfs(root.right, x + 1, y + 1);
  }
  
  dfs(root, 0, 0);
  const ySorted = [...yMap.keys()].sort((n1, n2) => {
    return n1 - n2;
  });

  const ans = [];
  for (const y of ySorted) {
    const vals = []
    xMap = yMap.get(y);
    const xSorted = [...xMap.keys()].sort((n1, n2) => {
      return n1 - n2;
    });
    for (const x of xSorted) {
      const xVals = xMap.get(x).sort((n1, n2) => {
        return n1 - n2;
      });
      vals.push(...xVals);
    }

    ans.push(vals);
  }

  // [1,2,3,4,5,6,7]
  // [1,2,3,4,6,5,7]
  return ans;
};
// @lc code=end

